modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php

```diff
@@ -0,0 +1,164 @@
+use Drupal\graphql\Plugin\GraphQL\SchemaExtension\SdlSchemaExtensionPluginBase;
+
+/**
+ * Adds topic data to the Open Social GraphQL API.
+ *
+ * @SchemaExtension(
+ *   id = "social_topic_schema_extension",
+ *   name = "Open Social - Topic Schema Extension",
+ *   description = "GraphQL schema extension for Open Social topic data.",
+ *   schema = "open_social"
+ * )
+ */
+class TopicSchemaExtension extends SdlSchemaExtensionPluginBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function registerResolvers(ResolverRegistryInterface $registry) {
+    $builder = new ResolverBuilder();
+
+    $this->addQueryFields($registry, $builder);
+    $this->addTopicFields($registry, $builder);
+  }
+
+  /**
+   * Registers type and field resolvers in the shared registry.
+   *
+   * @param \Drupal\graphql\GraphQL\ResolverRegistryInterface $registry
+   *   The resolver registry.
+   * @param \Drupal\graphql\GraphQL\ResolverBuilder $builder
+   *   The resolver builder.
+   */
+  protected function addTopicFields(ResolverRegistryInterface $registry, ResolverBuilder $builder) {
+    $registry->addFieldResolver('Topic', 'id',
+      $builder->produce('entity_uuid')
+        ->map('entity', $builder->fromParent())
+    );
+
+    $registry->addFieldResolver('Topic', 'title',
+      $builder->compose(
+        $builder->produce('entity_label')
+          ->map('entity', $builder->fromParent()),
+        $builder->produce('uppercase')
+          ->map('string', $builder->fromParent())
+      )
+    );
+
+    $registry->addFieldResolver('Topic', 'author',
+      $builder->produce('entity_owner')
+        ->map('entity', $builder->fromParent())
+    );
+
+    $registry->addFieldResolver('Topic', 'bodyHtml',
+      $builder->fromPath('entity:node', 'body.0.value')
+    );
+
+    $registry->addFieldResolver('Topic', 'heroImage',
+      $builder->produce('field')
+        ->map('entity', $builder->fromParent())
+        ->map('field', $builder->fromValue('field_topic_image'))
+    );
+
+    $registry->addFieldResolver('Topic', 'created',
+      $builder->fromPath('entity:node', 'created.value')
+    );
+
+    $registry->addFieldResolver('Topic', 'comments',
+      $builder->produce('query_comments')
+        ->map('parent', $builder->fromParent())
+        ->map('bundle', $builder->fromValue('comment'))
+        ->map('after', $builder->fromArgument('after'))
+        ->map('before', $builder->fromArgument('before'))
+        ->map('first', $builder->fromArgument('first'))
+        ->map('last', $builder->fromArgument('last'))
+        ->map('reverse', $builder->fromArgument('reverse'))
+        ->map('sortKey', $builder->fromArgument('sortKey'))
+    );
+
+    $registry->addFieldResolver('Topic', 'url',
+      $builder->compose(
+        $builder->produce('entity_url')
+          ->map('entity', $builder->fromParent()),
+        $builder->produce('url_path')
+          ->map('url', $builder->fromParent())
+      )
```

Comment:
I think `url_path` produces only the /path right? I think an absolute URL is expected here.

Reply:
Yeah, but we should create a custom Data Producer here I guess :(

Reply:
or we can do something like
```
    $registry->addFieldResolver('Topic', 'url',
      $builder->callback(function ($parent, $args) {
        return $parent->toUrl('canonical', ['absolute' => TRUE])->toString();
      })
    );
```

Reply:
There's nothing wrong with a custom data producer if it solves our problem and other data producers don't do that.

We can also take a closer look at the data producers we already have. Then we see `UrlPath` actually just turns a URL object into a string. So whether the returned URL is absolute depends on the configuration of the URL object passed to it.

Then we look at where that object comes from which is `EntityUrl` and we see it only does `$entity->toUrl()`. However, `EntityInterface::toUrl` does take options to that allow us to make the URL absolute. 

So rather than creating an entirely new DataProducer we can also make a contribution to the GraphQL module to provide optional `rel` and `options` arguments to the `entity_url` data producer. After we propose that as a PR to the GraphQL module we can add it to Open Social in two ways until it's merged.

1. We can make a patch and apply it to Open Social. This breaks as soon as the next minor version of the GraphQL module is released.
2. We can copy our PR into `social_core` as the `EntityUrl` class and give it the `entity_url_with_options` ID. Then add a comment to that class marking it as deprecated pointing to the PR in the GraphQL module. Once the GraphQL module's version is merged, we can update the code to use that class instead and then remove our own copy.

2 is only a little bit more work up-front, but is much less likely to break in the next weeks/months. So I think a contribution to the GraphQL module with option 2 to bring it into the distribution is desirable.

Reply:
So, I chose option 2 :) 
I have created PR to the `graphql` module, which provides the ability to get an absolute URL.

See https://github.com/drupal-graphql/graphql/issues/1219